File: cadwiki-nuget/cadwiki.AutoCAD2021.Base.Utilities/Shared/Commands.vb
Option Strict On
Option Infer Off
Option Explicit On

Imports System.Runtime.CompilerServices
Imports Autodesk.AutoCAD.ApplicationServices
Imports Autodesk.AutoCAD.DatabaseServices
Imports Autodesk.AutoCAD.EditorInput
Imports Autodesk.AutoCAD.Geometry
Imports Autodesk.AutoCAD.Runtime

Public Class Commands

    Public Shared Sub ExecuteInApplicationContext(parameters As List(Of Object))
        Dim dm As DocumentCollection = Core.Application.DocumentManager
        Dim doc As Document = dm.CurrentDocument
        dm.ExecuteInApplicationContext(
                Sub(obj)
                    doc.Editor.Command(parameters.ToArray())
                End Sub, Nothing)
    End Sub

    Public Shared Sub SendLispCommandStartUndoMark()
        Dim str As String = "(vla-startundomark (vla-get-ActiveDocument (vlax-get-acad-object)))"
        Core.Application.DocumentManager.MdiActiveDocument.SendStringToExecute(
            str + Constants.vbLf, True, False, False)
    End Sub

    Public Shared Sub SendLispCommandEndUndoMark()
        Dim str As String = "(vla-endundomark (vla-get-ActiveDocument (vlax-get-acad-object)))"
        Core.Application.DocumentManager.MdiActiveDocument.SendStringToExecute(
            str + Constants.vbLf, True, False, False)
    End Sub

    Public Shared Sub SendLispCommandUndoBack()
        Dim str As String = "(command-s ""._undo"" ""back"" ""yes"")"
        Core.Application.DocumentManager.MdiActiveDocument.SendStringToExecute(
            str + Constants.vbLf, True, False, False)
    End Sub
End Class

File: cadwiki-nuget/cadwiki.AutoCAD2021.Base.Utilities/Shared/Commands.cs
using System;
using System.Collections.Generic;
using Autodesk.AutoCAD.ApplicationServices;
using Autodesk.AutoCAD.Interop;
using Microsoft.VisualBasic;
using System.Threading.Tasks;

namespace cadwiki.AutoCAD2021.Base.Utilities
{

    public class Commands
    {

        public static void ExecuteInApplicationContext(List<object> parameters)
        {
            var dm = global::Autodesk.AutoCAD.ApplicationServices.Core.Application.DocumentManager;
            var doc = dm.CurrentDocument;
            dm.ExecuteInApplicationContext(obj => doc.Editor.Command(parameters.ToArray()), null);
        }

        public static void SendLispCommandStartUndoMark()
        {
            string str = "(vla-startundomark (vla-get-ActiveDocument (vlax-get-acad-object)))";
            global::Autodesk.AutoCAD.ApplicationServices.Core.Application.DocumentManager.MdiActiveDocument.SendStringToExecute(str + "\n", true, false, false);
        }

        public static void SendLispCommandEndUndoMark()
        {
            string str = "(vla-endundomark (vla-get-ActiveDocument (vlax-get-acad-object)))";
            global::Autodesk.AutoCAD.ApplicationServices.Core.Application.DocumentManager.MdiActiveDocument.SendStringToExecute(str + "\n", true, false, false);
        }

        public static void SendLispCommandUndoBack()
        {
            string str = "(command-s \"._undo\" \"back\" \"yes\")";
            global::Autodesk.AutoCAD.ApplicationServices.Core.Application.DocumentManager.MdiActiveDocument.SendStringToExecute(str + "\n", true, false, false);
        }

        public static void SendCommand(string command)
        {
            var acadApp = (AcadApplication)Application.AcadApplication;
            var thisDrawing = (AcadDocument)acadApp.ActiveDocument;
            thisDrawing.SendCommand(command);
        }

        public static async Task ExecuteInCommandContextAsync(Document doc, object[] parameters)
        {
            await global::Autodesk.AutoCAD.ApplicationServices.Core.Application.DocumentManager.ExecuteInCommandContextAsync(
                async (obj) =>
                {
                    await doc.Editor.CommandAsync(parameters);
                },
                null);
        }


        public static async Task RefreshRibbon(Document doc)
        {
            var refreshRibbonCommandLineArgs = new object[] { "RIBBON" };

            var myTask = Task.Run(() => ExecuteInCommandContextAsync(doc, refreshRibbonCommandLineArgs));
            await myTask;
            return;
        }

    }
}